### ISSUE TITLE
Get Post Objects by more than just the ID

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#207]
We should be able to get a single Post by more than just the ID. 

I'm thinking the postObjects should have a `by` input that allows you to do something like these: 

```
post( by: { slug: "the-post-slug" } ) {
   id
   title
}
```

```
post( by: { id: "opaqueId" } ) {
   id
   title
}
```

```
post( by: { postId: 123 } ) {
   id
   title
}
```

**Comment by CodeProKid** [comment]
Hmmmmm, I'm not sure I agree with this. I feel like if you want to get a post by anything other than the ID, it should be done with a `posts` query, since it could potentially return multiple posts in all instances, which makes the singular `post()` misleading.

**Comment by jasonbahl** [comment]
@CodeProKid so, the use case is that in a React (or other JS app), you will might be at a route, say: `site.com/post/post-slug` and you need to hydrate that route with that post. 

Much like WordPress itself `/site.com/post-slug`

So, you need to be able to fetch that resource, and the only identifier you might have for it is the slug. 

Since title's are _not_ unique, I can take that one out, but I think the postId (ID in the database), Opaque ID (base64 relay ID), and the slug, should be fields to get a post by.

**Comment by jasonbahl** [comment]
@CodeProKid I updated to remove the option to get by title.

**Comment by CodeProKid** [comment]
@jasonbahl you will still have the issue of possible name collisions though, since WordPress has a unique permalink structure. So, there could be a post in a custom post type called "My Post" that would have a slug of `my-post`, and the full permalink would be something like: `custom-post-type/my-post`. Then you could have another post in the Post post type called "My Post" that would have a slug of `my-post` and a full permalink like `blog/my-post` or whatever. So if you *only* passed the slug as an argument, you would get two posts back since they both have the same value in the `post_name` column. I think the only way we could do this is if we also required the Type as well.

**Comment by jasonbahl** [comment]
@CodeProKid ya, the Type is part of it anyway, as each post_type can only be queried via it's Type in the schema. 

like: 

```
post( by: { slug: "the-post-slug" } ) {
   id
   title
}
```

```
page( by: { slug: "the-post-slug" } ) {
   id
   title
}
```

**Comment by CodeProKid** [comment]
Ahhhhh, yeah you're right... I think I'm cool with this then. We _should_ still do some checks in the resolver to make sure it only returns one entity though, since I know we had some weird stuff happen with the direct DB migration with duplicate slugs within a post type causing an issue. Ultimately, I don't think it's safe to assume there won't be duplicate slugs in the DB.

**Comment by jasonbahl** [comment]
addressed by #208
